stake-pool/single-pool/tests/helpers/mod.rs

```diff
@@ -0,0 +1,423 @@
+        .expect("client error")
+        .expect("account not found")
+}
+
+// XXX using this unless i figure out how tf to get tarpc::context::Context
+#[allow(deprecated)]
+pub async fn get_fee_for_message(banks_client: &mut BanksClient, message: &Message) -> u64 {
+    let (fee_calculator, _, _) = banks_client.get_fees().await.unwrap();
+    fee_calculator.calculate_fee(message)
+}
+
+pub async fn create_vote(
+    banks_client: &mut BanksClient,
+    payer: &Keypair,
+    recent_blockhash: &Hash,
+    validator: &Keypair,
+    vote: &Keypair,
+) {
+    let rent = banks_client.get_rent().await.unwrap();
+    let rent_voter = rent.minimum_balance(VoteState::size_of());
+
+    let mut instructions = vec![system_instruction::create_account(
+        &payer.pubkey(),
+        &validator.pubkey(),
+        rent.minimum_balance(0),
+        0,
+        &system_program::id(),
+    )];
+    instructions.append(&mut vote_instruction::create_account(
+        &payer.pubkey(),
+        &vote.pubkey(),
+        &VoteInit {
+            node_pubkey: validator.pubkey(),
+            authorized_voter: validator.pubkey(),
+            authorized_withdrawer: validator.pubkey(),
+            ..VoteInit::default()
+        },
+        rent_voter,
+    ));
+
+    let transaction = Transaction::new_signed_with_payer(
+        &instructions,
+        Some(&payer.pubkey()),
+        &[validator, vote, payer],
+        *recent_blockhash,
+    );
+
+    // ignore errors for idempotency
+    let _ = banks_client.process_transaction(transaction).await;
+}
+
+pub async fn transfer(
+    banks_client: &mut BanksClient,
+    payer: &Keypair,
+    recent_blockhash: &Hash,
+    recipient: &Pubkey,
+    amount: u64,
+) {
+    let transaction = Transaction::new_signed_with_payer(
+        &[system_instruction::transfer(
+            &payer.pubkey(),
+            recipient,
+            amount,
+        )],
+        Some(&payer.pubkey()),
+        &[payer],
+        *recent_blockhash,
+    );
+    banks_client.process_transaction(transaction).await.unwrap();
+}
+
+pub fn check_error<T: Clone + std::fmt::Debug>(got: BanksClientError, expected: T)
+where
+    ProgramError: TryFrom<T>,
+{
+    // banks error -> transaction error -> instruction error -> program error
+    let got_p: ProgramError = if let TransactionError::InstructionError(_, e) = got.unwrap() {
+        e.try_into().unwrap()
+    } else {
+        panic!(
+            "couldnt convert {:?} to ProgramError (expected {:?})",
+            got, expected
+        );
+    };
+
+    // this silly thing is because we can guarantee From<T> has a Debug for T
+    // but TryFrom<T> produces Result<T, E> and E may not have Debug. so we cant call unwrap
+    // also we use TryFrom because we have to go `instruction error-> program error`
+    // because StakeError impls the former but not the latter...
+    // and that conversion is merely surjective........
+    // infomercial lady: "if only there were a better way!"
```

Comment:
Oof yeah, it's awful. If we need to add `Debug` to any sdk types to make this easier, I'm super game

Reply:
unless there is A Better Way, its not that simple... we simply dont have access to `E` in the context of the type signature because its not in the signature of `TryFrom<T>`. `TryFrom` itself would need to restrict its `E` to only those which impl `Debug`

instead i think we want to `impl From<StakeError> for ProgramError` but im not sure its worth touching the stake program to make some peripheral tests nicer... although maybe it would help someone else someday